sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -764,6 +773,94 @@ abstract class TypeCoercionBase {
     }
   }
 
+  object CollationTypeCasts extends TypeCoercionRule {
+    override def transform: PartialFunction[Expression, Expression] = {
+      case e if !e.childrenResolved => e
+
+      case b @ BinaryComparison(left, right) if shouldCast(Seq(left.dataType, right.dataType)) =>
+        val newChildren = collateToSingleType(Seq(left, right))
+        b.withNewChildren(newChildren)
+    }
+
+    def shouldCast(types: Seq[DataType]): Boolean = {
+      types.forall(_.isInstanceOf[StringType]) && types.distinct.length > 1
+    }
+
+    /**
+     *  Collates the input expressions to a single collation.
+     */
+    def collateToSingleType(exprs: Seq[Expression]): Seq[Expression] = {
+      val collationId = getOutputCollation(exprs)
+
+      exprs.map { expression =>
+        expression.dataType match {
+          case st: StringType if st.collationId == collationId =>
+            expression
+          case _: StringType =>
+            Cast(expression, StringType(collationId))
+        }
+      }
+    }
+
+    /**
+     * Based on the data types of the input expressions this method determines
+     * a collation type which the output will have.
+     */
+    def getOutputCollation(exprs: Seq[Expression], failOnIndeterminate: Boolean = true): Int = {
+      val explicitTypes = exprs.filter(hasExplicitCollation).map(_.dataType).distinct
+
+      explicitTypes.size match {
+        case 1 => explicitTypes.head.asInstanceOf[StringType].collationId
+        case size if size > 1 =>
+          throw QueryCompilationErrors
+            .explicitCollationMismatchError(
+              explicitTypes.map(t => t.asInstanceOf[StringType].typeName)
+            )
+        case _ =>
+          val dataTypes = exprs.map(_.dataType.asInstanceOf[StringType])
+
+          if (hasIndeterminate(dataTypes)) {
+            if (failOnIndeterminate) {
```

Comment:
also a sidenote about failOnIndeterminate; currently if will result sometimes in analysis and sometimes in runtime. Is this okay and should we maybe only fail in one place?